torch/export/export_schema.py

```diff
@@ -0,0 +1,319 @@
+#       These are faithful capture of tensor's detail in pytorch's executions during tracing
+#       However, it's up to downstream system on how to utilized these fields
+#       In another word, these feilds are suggestive, rather than mandatory.
+
+@dataclass
+class TensorMeta:
+    dtype: ScalarType
+    sizes: List[SymInt]
+
+    # needed for training
+    requires_grad: bool
+
+    # !!! see description above, there are subtle difference on how these fields should be interpreted
+    device: Device
+    strides: List[SymInt]
+    storage_offset: int
+    layout: Layout
+
+    # !!! memory_format is not tensor property, but it can only used as operator argument
+    # memory_format: MemoryFormat
+
+
+@dataclass
+class Buffer:
+    # !!! TODO: need to define endianness for the buffer
+    buffer: bytes
+
+
+@dataclass
+class ExternalBuffer:
+    # !!! Example from onnx
+    #   // Data can be stored inside the protobuf file using type-specific fields or raw_data.
+    #   // Alternatively, raw bytes data can be stored in an external file, using the external_data field.
+    #   // external_data stores key-value pairs describing data location. Recognized keys are:
+    #   // - "location" (required) - POSIX filesystem path relative to the directory where the ONNX
+    #   //                           protobuf model was stored
+    #   // - "offset" (optional) - position of byte at which stored data begins. Integer stored as string.
+    #   //                         Offset values SHOULD be multiples 4096 (page size) to enable mmap support.
+    #   // - "length" (optional) - number of bytes containing data. Integer stored as string.
+    #   // - "checksum" (optional) - SHA1 digest of file specified in under 'location' key.
+    #   repeated StringStringEntryProto external_data
+    location: str
+    offset: str     # !!! Consider using int, but int has int_max limitation
+    length: str     # !!! Consider using int, but int has int_max limitation
+    checksum: str
+
+
+@dataclass
+class Storage:
+    class DataLocation(Enum):
+        Internal = auto()
+        External = auto()
+
+    data_location: DataLocation
+    data: Union[Buffer, ExternalBuffer]
+
+
+# This is a concrete tensor backed by storage
+@dataclass
+class Tensor:
+    # storage
+    storage: Storage
+
+    # metadata
+    meta: TensorMeta
+
+
+################################################################################
+# Following section is the defining the schema of 3 level construct: GraphModule, Graph, Node
+
+# IValue has no corresponding class in fx
+# IValue is the "values" that are passed between nodes in the graph
+# !!! Assumption: Only Tensor can be passed between nodes, and not other int/float/bool types...
+# IValue is a named virtual tensor, with an optional TensorMeta that describes the properties of the tensor
+# !!! Consider using a more descriptive name, e.g. TensorValue, TensorPlaceholder, TensorArgument, etc.
+@dataclass
+class IValue:
```

Comment:
I want to propose a name change to the IValue class to `ValueDef` for a couple of reasons:
1. We already have an internal concept in pytorch core called `IValue` for the runtime, which might be confusing to some folks.
2. `Def` seems to better represent the idea of an identifier + metadata for the value.
3. `Def` is consistent with the def-use concept in traditional compiler infrastructures.

Reply:
@SherlockNoMad ^

Reply:
Agree that we should avoid IValue for confusion. I have no strong opinion on names  (as stated in the code comments). 
I will keep this comment open, and looking for better names.

Reply:
MLIR nomenclature is Result. Value is a pretty common naming for multi return IR too.

Reply:
> MLIR nomenclature is Result.

Yep, MLIR has OpResult which itself extends from Value. So it depends what context is this used for, but it sound that indeed we're talking about results here.